### ISSUE TITLE
Universal syncing library(ies)

### ISSUE THREAD

**Issue by jgaehring** [farmOS/farmOS#166]
I'd like to document an idea @mstenta and I discussed a few weeks ago regarding farmOS syncing libraries. I don't want to focus too much on the possible implementation, but rather the intention and utility of such a library, or libraries.

I think one of the main requirements that Mike and I identified for any application which wishes to sync data with a farmOS server is that it will need to, essentially, persist a local copy of the same data that's on the server, or some portion of that data; so in essence it will need to mirror the data schema of the farmOS server, at least in some respects. The first goal then of the syncing library would be to compose the data that comes from the server (via something like farmOS.js) into a workable schema that could easily be transferred to a storage medium. I think the crucial point here is that the library would leave the actual storage up to the user; the library would just be _preparing_ that data for storage. Therefore, an important question to ask when determining whether this project would prove useful or not is, "Will it prove useful and effective to prep such data beforehand in a separate process which can be abstracted into a library, or does it make better sense to leave that step entirely up to each potential user, let them decide how to structure the data as well as persist it?"

The second goal of this library would be structuring certain metadata around the farmOS entities (logs, assets and areas) for the purpose of comparing and reconciling changes that occur locally with changes that happen on the server. This would include such metadata as references to the data itself, timestamps of when each property on an entity has been changed locally, a local id for entities which have been generated locally but not yet assigned an id by the server, and, where necessary, a record of conflicts between the local copy and the server copy of an entity which cannot be automatically resolved but would require some kind of action from the end user to reconcile. This metadata would also need to be persisted, but again that would be up to the user.

Like the various networking libaries (eg, [farmOS.js](https://github.com/farmOS/farmOS.js), [farmOS.py](https://github.com/farmOS/farmOS.py), etc), we could potentially have implementations in different languages, and we could design the syncing libraries to work in close conjunction with those networking libraries. A potential third goal of the syncing libraries could be to provide a simple `.sync()` method, which calls methods of the networking library under the hood. In such a case, then, the corresponding networking library would be a dependency of the syncing library.

**Comment by gbathree** [comment]
+1 on having implementations in different languages, that's awesome.

I think the core issue is the versioning of changes and who overwrites who... I have a suggestion on that.  I think user roles in farmOS are there for exactly this purpose - they are defining who gets to make the call out of several people in a group.  I think in this case, both informing users about who's changed what when is going to get confusing for the user fast, and case more problems than it solves.  I think 98% of the time, the user role is telling you who should overwrite who.  If you're the farm manager, you should overwrite a farm worker, etc.  And it should be explicit in the description of user roles that you have that power, so they're not surprised when it happens.

Just a thought, but I am concerned about usability / complexity when it comes to syncing.It feels like as software engineers we can come up with systems which technically solve the problem of syncing, but make it nearly impossible to provide a reasonable user experience.

Them's my thoughts!

**Comment by mstenta** [comment]
Thanks for the thoughts @gbathree !

> I think the core issue is the versioning of changes and who overwrites who...

Actually I think this is a separate issue from the overall "universal syncing libraries". Jamie started one in the farmOS-client queue: https://github.com/farmOS/farmOS-client/issues/190

But on that topic...

> I think user roles in farmOS are there for exactly this purpose - they are defining who gets to make the call out of several people in a group. I think in this case, both informing users about who's changed what when is going to get confusing for the user fast, and case more problems than it solves. I think 98% of the time, the user role is telling you who should overwrite who. If you're the farm manager, you should overwrite a farm worker, etc. And it should be explicit in the description of user roles that you have that power, so they're not surprised when it happens.

This sounds like a good idea, and I'm all for keeping things simple, but I don't think leaving it as a role-level automated decision will work. The core issue is data loss - and preventing that - which requires communicating with the user and giving them choices. I don't think there's any way around that.

Consider this scenario: you have two people on the farm, both with the "Farm Manager" role. User 1 creates a todo list for the day, and User 2 syncs that to their phone. User 2 goes out into the field to perform tasks, and no longer has internet access. One of the todo items is to harvest a crop - and the log is already generated, but is marked as "not done" and the "quantity" field is left blank (to be filled in by User 2 when they are done harvesting). While they are harvesting, User 1 updates the log and adds some information to the "Notes" field (simple example). Meanwhile, User 2 is harvesting, and ALSO adds some notes to the log. Now the two logs are no longer in sync. If we just let User 2's changes overwrite User 1's, then User 1's notes are lost - and more importantly: **no one knows that they were lost**.

Granted, this probably won't be a common issue. But it is something that the farmOS-client app needs to be able to intelligently resolve. Jamie has a plan for creating a UI that shows the difference between the server and client versions of a log: https://github.com/farmOS/farmOS-client/issues/190

Looking ahead... in Drupal 8 we will be adding the ability to save revisions on all records in farmOS. Which means, even if User 2 overwrites User 1's changes, there will be a revision saved for each - so the data won't be completely lost.  At that point, if we detected a conflict, we could allow it to take place, but also add a "Needs Review" flag (or something) to the log, so that it can be reviewed later, and adjusted if necessary.

**Comment by gbathree** [comment]
Ok - fair enough, I agree that if you can do versioning well enough (technically) and the interface easy enough (UX) that's ideal.  I'm just worried that that is harder than it seems, but I may be a cynic :) 

Perhaps also you could consider identifying where these edge cases will occur (or identifying via user testing even better if you can implement with actual farms) and then address in a specific way, rather than trying to make a more complex, generalized UX for displaying/merging conflicts.

But on net, I hear you, and it sounds like an awesome solution if done well!

**Comment by jgaehring** [comment]
I guess I'm thinking more in terms of DX than UX; that is, I'm trying to defer a lot of the UX choices to the developer who would be using the library. Therefore my focus is on the underlying data model for versioning changes and the core logic for syncing changes and resolving conflicts, without necessarily making assumptions how the developer using this library would prioritize one change over another.

I should preface this by saying that I'm imagining something like a JavaScript state tree, like you would use in Redux or Vue, but I'm hopeful the idea translates to other languages and idioms as well.

Right now, I'm envisioning 3 modes of representing farmOS data, which would represent the first 3 branches of the state tree: the local data itself, its metadata, and remote conflicts.

```js
data: {
  // The data would be as close as possible to the actual request/response data
  // that is sent to and received from the server.
  logs: [
    {
      id: 205,
      // Probably the one difference between local data and server data is that
      // there would have to be a local id (lid) assigned, for tracking locally 
      // generated logs, areas, etc.
      lid: 5,
      notes: 'We were only able to harvest 8 bushels because of mildew'
      // ... all the rest of the log data
    },
    // more logs
  ],
  areas: [ /**/ ],
  assets: [ /**/ ],
  // etc
},
// I guess this could just as easily be called "changes" or something
metadata: {
  logs: [
    {
      id: 205,
      lid: 5,
      // Apart from the lid and id, all other metadata would be timestamps for
      // when each field was last changed locally.
      notes: 1556571213,
    },
    // more logs
  ],
  areas: [ /* all the area metadata */ ],
  assets: [ /* all the asset metadata*/ ],
  // etc
},
conflicts: {
  // These would represent the logs, assets, etc as they exist on a remote
  // device (presumably the server), which have conflicts that cannot be 
  // resolved automatically. These would only contain those fields which were
  // found to be conflicting, so they would not be complete logs/etc as above
  // in the data section.
  logs: [
    {
      resolved: false,
      lid: 5,
      notes: 'Harvest 10 bushels and check for signs of downy mildew'
    },
    // more logs
  ],
  areas: [ /* all the area conflicts */ ],
  assets: [ /* all the asset conflicts */ ],
  // etc
}
```

I'm sure there are other ways of structuring it, and other features we could provide (eg, import filters, time of last sync action, etc), but this would basically preserve all the data as thoroughly but as concisely as possible. That way, the developer wouldn't have to worry about the details, which would be handled by private methods, but they could call some public methods for retrieving that data and making their own decisions about how to resolve conflicts etc, or provide a UI and have user decide. For instance, in Field Kit, we're basically planning to display those remote conflicts to the user with a choice of which to pick, and possibly the option to overwrite with their own input (so they can manually merge if needed).

So the main goal of the syncing library would be to resolve all changes that can be resolved uncontroversially (what that constitutes could be problematic, admittedly), and then to surface version history and unresolved conflicts in an organized fashion that the developer could then process more efficiently and easily persist to disc.

**Comment by jgaehring** [comment]
Wow, just found this, could be useful: https://github.com/automerge/automerge

Via https://medium.com/@pvh/pixelpusher-real-time-peer-to-peer-collaboration-with-react-7c7bc8ecbf74, via https://www.inkandswitch.com/local-first.html#future

**Comment by jgaehring** [comment]
So in the course of trying to grok all this CRDT stuff, and also trying to evaluate if any of it could be useful for the synchonization process in Field Kit, I've started to wonder if we do need some changes to how data is structured on the server to support concurrency with many offline devices, each with their own replicated copy of the farmOS data. The `changed` timestamp helps a lot, but it's not very fine-grained control. I was particularly struck by the way one of the CRDT's used autogenerated UUID's for each field in JSON data, and I wonder if something like this, implemented on both the server and the client, could help. Or perhaps something much simpler which just records some more fine-grained version history. @mstenta has mentioned the that farmOS could enable "entity revisions", which is another concept I'm unfamiliar with, but sounds promising.

In any event, I'm realizing more and more that if we're really committed to offline-first design with Field Kit, and we want to support some of the use cases where the farm worker's copy of the data needs to take priority over the farm manager's, then we really can't treat Field Kit's local data as strictly ephemeral, nor the server's as always canonical. If we do, we will always risk data loss. Perhaps a certain degree of data loss is acceptable, but personally I'd like to aim to eliminate that risk entirely. And to do so I think we'll have to acknowledge the concurrent nature of the problem we're trying to solve, when we introduce offline-first clients to the larger ecosystem of farmOS.

I'm not sure best way to proceed from there, because concurrency and distributed systems are totally new to me, but I'm curious what other people's thoughts are on this. Is it worth exploring further in this issue thread, or perhaps elsewhere? Or am I just blowing this all out of proportion? CRDT's are probably overkill for our scenario, but what other options are out there?

**Comment by gbathree** [comment]
It ain't easy, that's for sure... I don't have great thoughts, but I think it's either a roll your own and keep it simple solution, or follow someone method who's really thought through the details (which you've found it seems).   Rolling your own to the degree you're creating a full blown, offline first, full choice solution feels like you're taking on a lot... but you're deeper into this than me for sure.

Basically, ¯\_(ツ)_/¯

**Comment by jgaehring** [comment]
> but I think it's either a roll your own and keep it simple solution, or follow someone method who's really thought through the details

I think you're right @gbathree , and I'd probably opt for the former, because at the end of the day we're not doing anything so terribly complicated as trying to create a Google Docs clone with real-time collaboration; we have the luxury of working with simple forms which can easily be represented as JSON, and what few conflicts arise should be rather straight-forward to merge (or to record and dispatch to the UI).

This might be a question for @mstenta when he's back from vacation, but my one lingering concern is: even for a simple solution, do we need to have an identical data structure for tracking changes on the server in order for it to avoid data loss? Even if it's as simple as a timestamp for each field, kind of like we're doing now on the client (but not yet on the server)?

**Comment by mstenta** [comment]
> do we need to have an identical data structure for tracking changes on the server in order for it to avoid data loss

We will be limited to what Drupal provides us, in this regard. But with Drupal entity revisions (not yet implemented on logs, assets, or taxonomy terms), it should have all the info we need - albeit not in the same format as you're describing.

Revisions in Drupal basically store a copy of each version of an entity in the database. So if you change one field, a new "revision" of the entity is created, along with a timestamp and user ID to indicate who performed the change and when.

So in theory, you could construct the information you're describing (field-level timestamps) from that. Although it wouldn't come out-of-the-box from the Drupal REST API module like that. So some reformatting would be required in the JS code (or we would need to expose a custom API endpoint in farmOS that formats it in that way - which I would be reluctant to do).

Two other challenges/questions/considerations:

1. Does Drupal's REST API module expose revisions? Or does it only show the most recent version of an entity?
2. I wasn't planning on implementing entity revisions until Drupal 8. :-/

**Comment by mstenta** [comment]
> the information you're describing (field-level timestamps)

Maybe we can have a more in-depth call/chat about these ideas next week, if it's pressing.

**Comment by jgaehring** [comment]
> if it's pressing

Not at all, but having spent quite a bit of time with this issue now, it brought up some questions I thought it would be good to consider for long-term development. I just wanted to plant the seed, so if in the future concurrency across devices becomes more of a serious issue, we're not caught totally flat-footed.

**Comment by mstenta** [comment]
Agreed! Thanks for getting these thoughts documented! I think this is something that will become more necessary in the greater Drupal community as well, so it would be interesting to do some research into similar projects/implementations to see if anyone else has given it thought.

**Comment by jgaehring** [comment]
I should also note that I've been working much more closely with the implementation we have in Field Kit and feel more confident that the essential design should work well enough to prevent data loss once we iron out some of the kinks.

**Comment by jgaehring** [comment]
Thinking about this again. The current state of Field Kit's [`farmLog`](https://github.com/farmOS/farmOS-client/blob/v0.6.0/src/utils/farmLog.js) puts us _really_ close to this actually achieving this. I just need to prune some of the non-essential metadata that's specific to FK and we'd have a JavaScript lib that anyone could use for keeping logs in sync with a farmOS server.

The real kicker would be if, with farmOS 2.x, we could generate our own UUID. Then we'd no longer need to keep track of our own localID. @mstenta, will that be possible?

**Comment by gbathree** [comment]
I would love a download on how this works... I think Manuel is using it in the integration, but unclear.  We spent a lot of time using basic JS to compose API calls to send info to farmOS, but I'd be curious to see if this augments or replaces what we do now.  I'd love to be a fly on the wall if there's a discussion about it.

**Comment by jgaehring** [comment]
I'd love to upload! 

TL;DR: [farmOS.js](https://github.com/farmOS/farmOS.js) is still our networking library, but internal to Field Kit I've been developing `farmLog`, which could become a part of that library, or it's own separate library. It's basically a set of functions for processing logs, whether it's a log you're creating locally, one you've pulled from the server, etc.


Here's how you might use it in conjunction with farmOS.js:

```js
import farmOS from 'farmos';
import farmLog from 'farmLog';

const farm = farmOS('example.farmos.net', { clientId: 'farm_client' });
let logs, localLog;

farm.authorize('bob@farmail.com', 'password')
  .then(() => farm.info())
  .then((res) => {
    const types = res.resources.log;
    logs = farmLog(types);
    const { createLog, formatLogForServer } = logs;
    localLog = createLog({
      name: 'My brand new log!',
      type: 'farm_observation',
      localID: 1,
    });
    return farm.log.send(formatLogForServer(localLog));
  })
  // Meanwhile, the server changes the name of the log...
  .then(({ id }) => {
    const { updateLog } = logs;
    localLog = updateLog(localLog, {
      id,
      name: 'Update from the client',
    });
    return farm.log.get(id);
  })
  .then((res) => {
    const { mergeLogFromServer } = logs;
    const serverLog = res.list[0];
    localLog = mergeLogFromServer(serverLog, localLog);
    const { conflicts } = localLog.name;
    console.log(conflicts);
    // PRINTS:
    // [
    //   {
    //     changed: 1596737193,
    //     data: 'Update from the server!',
    //   },
    //   {
    //     changed: 1596737194,
    //     data: 'Update from the client!',
    //   }
    // ]
  })
```

**Comment by mstenta** [comment]
@gbathree I don't think you are using `farmOS.js` on your server (unless that has changed in the PWA and you are running node.js now?) Worth asking Manuel perhaps. But... this may still be useful... if it allows you to more easily compose log JSON in the frontend, and then send that to the backend to be sent to farmOS by the SurveyStack server.

Also note that right now the stuff @jgaehring is talking about is ONLY for logs... you are also creating assets and terms.  But once we have the logs nailed down in farmOS.js it makes sense to add similar support for the other record types as well. So yes! I think this could be really helpful for script composing in SurveyStack in the future!

**Comment by mstenta** [comment]
@jgaehring - Keep that in mind... SurveyStack is NOT authenticating/sending requests directly from the client. It has to go through their server so that the API key is not exposed on the client.

So... it would be great if they could load farmOS.js but only use it to "build" log records... without using any of the actual request/sync code, if that makes sense.  Building the log JSON is the main requirement.

**Comment by jgaehring** [comment]
> So... it would be great if they could load farmOS.js but only use it to "build" log records... without using any of the actual request/sync code, if that makes sense. Building the log JSON is the main requirement.

Yea, `farmLog` is totally independent of farmOS.js, at least at the moment, so that's probably a good argument in favor of keeping it that way (ie, not integrating it with farmOS.js).

**Comment by mstenta** [comment]
I'm going through and closing old/stale issues and discussions. Moving forward we are encouraging general discussions take place on the [farmOS community forum](https://farmOS.discourse.group). GitHub will be reserved for bug reports and pull requests.

@jgaehring @gbathree feel free to open a new forum topic (or use an existing one) and link to this if you want!
